src/main/java/org/jenkinsci/remoting/protocol/IOHub.java

```diff
@@ -489,6 +514,44 @@ public final void run() {
             // ignore, happens routinely
         } finally {
             selectorThread.setName(oldName);
+            ioHubRunning = false;
+        }
+    }
+
+    /**
+     * This is an artificial thread, which monitors IOHub Selector and wakes it up if it waits for more than 1 second.
+     * It is a workaround for Selector#select(long timeout) on Windows, where the call always waits for the entire timeout before returning back.
+     */
+    private static class WindowsIOHubSelectorWatcher implements Runnable {
+
+        private final IOHub iohub;
+
+        public WindowsIOHubSelectorWatcher(IOHub iohub) {
+            this.iohub = iohub;
+        }
+
+        @Override
+        public void run() {
+            final Thread watcherThread = Thread.currentThread();
+            final String oldName = watcherThread.getName();
+            final String watcherName = "Windows IOHub Watcher for " + iohub.getThreadNameBase(oldName);
+            LOGGER.log(Level.FINEST, "{0}: Started", watcherName);
+            try {
+                watcherThread.setName(watcherName);
+                while (iohub.ioHubRunning) {
+                    Thread.sleep(1000);
```

Comment:
:bug: causes the executor service to excessively wait on shutdown. Better to use a lock object for running and use `wait/notifyAll`